certificate/certificates.go

```diff
@@ -58,6 +58,15 @@ type ObtainRequest struct {
 	PreferredChain string
 }
 
+// ObtainForCSRRequest The request to obtain a certificate matching the CSR passed into it.
+//
+// If bundle is true, the []byte contains both the issuer certificate and your issued certificate as a bundle.
+type ObtainForCSRRequest struct {
+	CSR            x509.CertificateRequest
```

Comment:
I thought about this when I made the structure.
Since this is a mandatory parameter I liked the idea of using a structure rather than a pointer.
But I am open to discussion.

Reply:
Oh, I'm all for an argument struct, since Go doesn't to named arguments :-)

I would like to change the type of `ObtainForCSRRequest.CSR` from `x509.CertificateRequest` to `*x509.CertificateRequest` though. Looking at the call-sites, `ObtainForCSRRequest` is always built like this:

```go
// This is from (c *Certifier).Renew(), but code in cmd/cmd_renew.go,
// cmd/cmd_run.go and e2e/challenges_test.go is very similar. All are
// dereferencing a csr pointer:
c.ObtainForCSR(ObtainForCSRRequest{
	CSR: *csr,
	// ...
})
```

This is because `x509.ParseCertificateRequest()` already returns a pointer.

In `ObtainForCSR()`, we'll pass the pointer along:

```go
domains := certcrypto.ExtractDomainsCSR(&request.CSR)
cert.CSR = certcrypto.PEMEncode(&request.CSR)
```

Reply:
maybe my comment was ambiguous, but I also talk about the `CSR` field

Reply:
Yeah, that's what I meant with "It would however require (*Certifier).ObtainForCSR() to check if request.CSR is nil (and bail with an error)". Something like:

```go
var ErrObtainMissingCSR = errors.New("cannot obtain resource for CSR: CSR is missing")

func (c *Certifier) ObtainForCSR(request ObtainForCSRRequest) (*Resource, error) {
	if request.CSR == nil {
		return nil, ErrObtainMissingCSR
	}

	// figure out what domains it concerns 
	// ...
}
```

Reply:
So changed the type